Pvl Core Addition

acpaquette opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR adds a long standing effort to extract the core file reading and writing components from the main `src` files in ISIS into there own, individually buildable library. All files that were moved should match the files in dev and should just be moved into a new folder.

## Related Issue
N/A

## How Has This Been Validated?
Validated through passing Jenkins tests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Infrastructure change (changes to things like CI or the build system that do not impact users)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

AustinSanders committed:
Added history comment.

Commit:
Initial move to core

Commit:
Removed Pvl cmakelists

Commit:
Removed more cmakelists.txt

Commit:
Add subdirectory

AustinSanders committed:
Fully modularized core

AustinSanders committed:
Fully modularized core

AustinSanders committed:
replaced PvlEditDialog that was accidentally removed in prev commit

AustinSanders committed:
Initial constants unit gtest

jessemapel committed:
Fixed pvl core test config

jessemapel committed:
Changed cmake_source to project_source for submodule

AustinSanders committed:
First working swig wrappers

jessemapel committed:
Fixed pvl core test config (#4349)

* Fixed pvl core test config

* Changed cmake_source to project_source for submodule

AustinSanders committed:
Rebased off of pvl_core

jessemapel committed:
Moved old ISIS unitTests into core (#4351)

* First pass at old unitTests

* re-removed old files

* Slim down and unittest port

* installing IsisPreferences

* Fixed tests in a full build

acpaquette committed:
Finished swig wrapper with some syntax sugar for ease of use in python (#4352)

AustinSanders committed:
Initial core deployment (#4355)

jessemapel committed:
Updated preference files (#4354)

acpaquette committed:
Reset main build and created core recipe

Forgot meta.yaml

Change where the build is pulling from

Fixed qt requirement in core recipe

Removed complier definition

Turn swig build on

acpaquette committed:
Changed pvl recipe url to point to usgsastro

acpaquette committed:
Fixed typo

acpaquette committed:
Removed a chunk of files that made it through the rebase (#4523)

Commit:
Add Blob to core from base (#4525)

* Fixed tests

* Remove Blob from base

acpaquette committed:
Pvl core Docs (#4536)

* Small doc string changes

* Fixed more doc strings and updated RTD configs

* Added core docs

* Removed path setup from config

* Added subprocess import

* Added back os import

* Added breathe to env file

* Maybe fix?

* Added doxyfile

* Trying doxygen change

acpaquette committed:
Blob Swig Wrapper (#4613)

* Moved wrapper under one package

* Add pvl construction from reading a string, give access to blob internals

* Updated meta.yaml to build isisio rather than isispvl

acpaquette committed:
Post rebase fixes

acpaquette committed:
Fixed app logging for all call able apps

acpaquette committed:
Moved table reading/writing into core

acpaquette committed:
Fixed missing license identifiers, and leftover merge conflict from rebase

acpaquette committed:
Exposed error formatting and file line reporting flags on preference class

acpaquette committed:
Updates IException to use the exposed Preference flags

acpaquette committed:
Application handles adding program name to error

acpaquette committed:
Removed unnecessary includes to classes outside of core

acpaquette committed:
Removed tempFixtures include in core tests

acpaquette committed:
Updates sources and includes for building

acpaquette committed:
Moves necessary fixtures into core

acpaquette committed:
Enforce that core unit tests use the test preferences

acpaquette committed:
Removed commented Application::Name all

acpaquette committed:
Initial round of test fixes

acpaquette committed:
Fixed last 3 failing gtests

acpaquette committed:
Fixed AppendAndLog function uses

acpaquette committed:
Updated IException test truth data

acpaquette committed:
Caught another missused AppendAndLog call

acpaquette committed:
Fixed last failing gtest

acpaquette committed:
Fixed output from getsn

acpaquette committed:
Moved Endian into core

acpaquette committed:
Remove additional 'm' extension for python 3.10

acpaquette committed:
Set version to 0.0.1 and clamped qt bellow 5.15

acpaquette committed:
Initial changes for isis pvl python interface

acpaquette committed:
Fixed missed conflict in rebase

acpaquette committed:
Removed prints causing test failures

Kelvinrr commented on isis/src/base/objs/Application/Application.cpp:
```diff
@@ -387,6 +382,21 @@ namespace Isis {
+   * @param log App log for running in code
+   *
+   */
+  void Application::AppendAndLog(PvlGroup &results, Pvl *log) {
+    Application::Log(results);
+    if (log) {
+      log->addGroup(results);
+    }
+  }
```
Oof, I forgot about this one. Would this constitute an API breaking change? 🤔

acpaquette replied:
I don't think so as I didn't remove any functionality but it is an addition. Likely an 8.X.0 release

jlaura commented on isis/src/core/CMakeLists.txt:
```diff
@@ -0,0 +1,113 @@
+cmake_minimum_required(VERSION 3.10)
+
+project (core VERSION 0.0.1)
```
Is PVL really a 0.0.1 version? That would be like a just developed, alpha, probably do not use this because nothing is stable version. What are the criteria for this to be a 1.0.0 release that folks should be comfortable using?

Kelvinrr replied:
I think we agreed a while back that if it's going public it should be 1.0, So I imagine this should also be 1.0? We plan to do a conda deploy after this is merged.

acpaquette replied:
Agree with both, I just used 0.0.1 as a place holder

acpaquette replied:
Updated to 1.0.0

Kelvinrr commented on rtd_docs/conf.py:
```diff
@@ -49,4 +49,18 @@
+
+# -- Run Doxygen for C++ API --------------------------------------------------
+
+read_the_docs_build = os.environ.get('READTHEDOCS', None) == 'True'
+
+if read_the_docs_build:
+    subprocess.call('doxygen Doxyfile', shell=True)
+
+# -- Breathe Configuration ----------------------------------------------------
+
+breathe_projects = { "ISIS": "doxyxml/" }
+breathe_default_project = "ISIS"
+
+breathe_domain_by_extension = {"h" : "cpp"}
```
Is this related to the core stuff? Seems like an odd inclusion. Im guessing it's for building the PVL core with docs. If so we might want to add a new objective of getting the docs pushed to the S3 buckets and the landing page, but that can happen after this is merged.

amystamile-usgs commented:
I'm sure you are already aware of these test failures:

	2077 - Mosrange.MosrangeDefault (Failed)
	2078 - Mosrange.MosrangeOnErrorContinue (Failed)
	2079 - Mosrange.MosrangeOnErrorFail (Failed)

Kelvinrr commented:
> I'm sure you are already aware of these test failures:
> 
> ```
> 2077 - Mosrange.MosrangeDefault (Failed)
> 2078 - Mosrange.MosrangeOnErrorContinue (Failed)
> 2079 - Mosrange.MosrangeOnErrorFail (Failed)
> ```

These may be hold overs from a previous PR that fixed the errors. We probably need updated tests.

acpaquette committed:
Update version to 1.0.0

acpaquette committed:
Added cxx-compiler as a build dependency

acpaquette committed:
Added mesalib as a build and run dependency

acpaquette committed:
Changed all Endian.h to IEndian.h to prevent include shadowing in a case-folding file system

acpaquette committed:
Fixed python find for python 3.9

Kelvinrr commented:
hmmm, those tests are still failing so maybe they aren't from the previous PR.

KrisBecker commented:
> hmmm, those tests are still failing so maybe they aren't from the previous PR.

The `pvl_core` branch is one commit behind `dev`. That commit fixes these test failures. You need to merge `dev` into the `pvl_core` branch to fix it.

acpaquette committed:
Fixed mosrange test failures

acpaquette commented:
@KrisBecker I rebased off of dev and the issue was still present. There is an extra function to both log to an open application window and add to the pvl log output. I have fixed the issue and this should be good to merge

acpaquette committed:
Fix python find for 3.11

acpaquette commented:
@KrisBecker To address some of the API breaking issues. I had to move the logging from `PvlGroup` which accessed the `Application` class. This was not an ideal coupling. I changed this so that the `Application` class would log to both the application and the resulting pvl output to maintain current functionality. 

The resulting test failures were from a poor git rebase on my part. Regardless, there is an API breakage under the `PvlObject` where the `appendLogGroup` function has been stripped of any functionality. This was done after the decouple. Unfortunately with the semantic versioning, this PR will have to wait a year for that function to to deprecated and the functionality to be added under 9.0. I will close this PR and start working on getting the application log changes into its own PR.
